src/OneDriveSdk/Helpers/UploadSessionHelper.cs

```diff
@@ -0,0 +1,257 @@
+        /// <param name="maxChunkSize">Max size of each chunk to be uploaded. If less than 0, default value of 10 MB is used. Increment of 320 kb is
+        /// recommended.</param>
+        public UploadSessionHelper(UploadSession session, IBaseClient client, Stream uploadStream, int totalUploadLength, int maxChunkSize = -1)
+        {
+            if (!uploadStream.CanRead || !uploadStream.CanSeek)
+            {
+                throw new ArgumentException("Must provide stream that can read and seek");
+            }
+
+            this.Session = session;
+            this.client = client;
+            this.uploadStream = uploadStream;
+            this.totalUploadLength = totalUploadLength;
+            this.rangesRemaining = this.GetRangesRemaining(session);
+            this.maxChunkSize = maxChunkSize < 0 ? DefaultMaxChunkSize : maxChunkSize;
+        }
+
+        /// <summary>
+        /// Get the series of requests needed to complete the upload session. Call <see cref="UpdateSessionStatusAsync"/>
+        /// first to update the internal session information.
+        /// </summary>
+        /// <param name="options">Options to be applied to each request.</param>
+        /// <returns>All requests currently needed to complete the upload session.</returns>
+        public IEnumerable<UploadChunkRequest> GetUploadChunkRequests(IEnumerable<Option> options = null)
+        {
+            foreach (var range in this.rangesRemaining)
+            {
+                var currentRangeBegins = range.Item1;
+
+                while (currentRangeBegins <= range.Item2)
+                {
+                    var nextChunkSize = NextChunkSize(currentRangeBegins, range.Item2);
+                    var uploadRequest = new UploadChunkRequest(
+                        this.Session.UploadUrl,
+                        this.client,
+                        options,
+                        currentRangeBegins,
+                        currentRangeBegins + nextChunkSize - 1,
+                        this.totalUploadLength);
+                    
+                    yield return uploadRequest;
+
+                    currentRangeBegins += nextChunkSize;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the status of the session. Stores returned session internally.
+        /// Updates internal list of ranges remaining to be uploaded (according to the server).
+        /// </summary>
+        /// <returns>UploadSession returned by the server.</returns>
+        public async Task<UploadSession> UpdateSessionStatusAsync()
+        {
+            var request = new UploadSessionRequest(this.Session, this.client, null);
+            var newSession = await request.GetAsync();
+            
+            var newRangesRemaining = this.GetRangesRemaining(newSession);
+
+            this.rangesRemaining = newRangesRemaining;
+            newSession.UploadUrl = this.Session.UploadUrl; // Sometimes the UploadUrl is not returned
```

Comment:
>// Sometimes the UploadUrl is not returned [](start = 59, length = 42)

Does those comment imply we callers should be careful for null checks, or something we need to fix before pushing this change?